### DISCUSSION TITLE
problems with installing win 10 22h2 onto iscsi disk

### DISCUSSION THREAD

**Discussion by firemeteorxx** [ipxe/ipxe#817] [General]
I know this is not necessarily specific to ipxe. Just wanted to seek for advice...

Basically I can't get win 10 22h2 working on iscsi disk. For a fresh install, it reboots in the middle of the 1st stage and can't even proceed to the 2nd stage. The reboot happens roughly after the 1st step (copying files) is done but before any configuration.

The same ipxe + wds setup works just fine with win 7 or an ancient win 10 1709 version. The problem for 1709 is that it fails to upgrade to 2022 through windows upgrade. The upgrade will lead to an un bootable system and finally a rollback with code 0xc1900101. I have no idea if this is due to the same reason that fails the fresh install, but anyway I can't get a smooth upgrade this way.

Since I saw records saying that win 10 1903 is known to fail iscsi install, I wonder if 22h2 has similar known issue and / or known workaround? 

It will be very annoy if the system will forcedly upgrade into a broken new release and break existing working system. Win 7 is much better in this regard.

Ref:
http://mistyprojects.co.uk/documents/iscsi-targets/files/supported.htm

**Comment by NiKiZe** [comment]
None of this post will give solutions, only some background on what you might want to look for.

Windows setup usually have at least 2 reboots. Are you sure it dies in the first stage?, because that is purely winpe.

Some (many/most) NIC drivers do not support setting "start on boot" or boot critical, which also makes startup fail as soon as it hands over from iPXE driver to the winnt kernel.
One other well documented issue is the wpf(?) Filtering settings.

I assume your upgrade path is seeing similar issues with the NIC driver not being configured as boot critical, or incorrect "filter" settings.

There is several posts in discussions about iSCSI boot. Have you looked at them?

**Comment by firemeteorxx** [comment]
Thanks for your quick response, NiKiZe!

> Windows setup usually have at least 2 reboots. Are you sure it dies in the first stage?, because that is purely winpe.
I think I'm pretty sure it's an abnormal reboot. The installer haven't got any chance to setup the target disk. It's almost an immediate reboot after selecting the target partition, the disk is almost empty and with no boot-loader installed at all.

I reproduced the same syndrome on three different setups:
1. Virtual Box with 82540EM emulated NIC
2. NUC6 with Intel i219 NIC
3. DIY box with Intel i218 NIC

Both #2 / #3 has working iSCSI booted Win7 && Win10 1709 OS build. It's just 22H2 causing a lot of troubles.

> Some (many/most) NIC drivers do not support setting "start on boot" or boot critical, which also makes startup fail as soon as it hands over from iPXE driver to the winnt kernel.

I can't rule out this possibility. But on the other hand, I'm not sure how to check this out.
Should the chance be low given the reputation of these builds on the past?
But anyway if you can point me to any reference it will be highly appreciated.

> One other well documented issue is the wpf(?) Filtering settings.

This is the source of most of my pain when setting up my Win7 iSCSI box several years ago. So yes, I'm aware of this issue. 
But to be honest, I'm not clear if iSCSI boot should immune to Windows update.
But I would say Microsoft sucks if it isn't the case.

I attached the failing disk to another working Win7 box and manually examined the Registry.
With the help from a tool called bindview, I tentatively believe that the WFPLWF filter is not bound to my booting NIC.
But once again, this is not a rock solid conclusion -- I can easily make mistake...
Or maybe I was not in the best timing to investigate -- the Win10 box is able to roll itself back into 1709 after failing the upgrade and I couldn't remember at which stage I attached the disk for investigation.
I wonder how it achieved such rollback action if the NIC is not working properly?

Putting the debugging && fixing effort aside, I'm curious if other Win10 iSCSI boot users also suffer from such upgrade failure when Microsoft roll-out new upgrades? Even if we can fix it each time it would be pretty annoy, right? 
It really does not sound right to me. Or probably I have some issue on my side, just that I didn't figure out yet?

**Reply by NiKiZe** [reply]
If your winpe reboots early and on different hardware, then assume driver issues. I have learnt (the hard way)bto not have any modifications to winpe is the safest, then only load correct drivers as needed. Don't embed NIC or disk drivers, instead drvload by script the NIC that is needed. Then let the setup part search for any disk drivers. Rule out this by using minimal winpe and manually load drivers as needed. If that fails as well then tcpdump the iSCSI connection.

För all the other parts I'd start of by assuming that the NIC drivers used is not "iSCSI certified" and in your upgrade path I would try to ensure that the drivers is first upgraded, and that the setup does search for upgraded drivers during the upgrade process.

**Comment by firemeteorxx** [comment]
> I have learnt (the hard way)bto not have any modifications to winpe is the safest, then only load correct drivers as needed. Don't embed NIC or disk drivers, instead drvload by script the NIC that is needed.

The WinPE system from the Win10 22H2 ISO appears to recognize my NIC and disk just fine. So I didn't make any change to the install media... 

The stock driver from 22H2 ISO may be causing trouble in my case. Could you elaborate a little bit about the 'drvload by script' path so that I can try it out?

**Comment by firemeteorxx** [comment]
> If your winpe reboots early and on different hardware, then assume driver issues.

I have to clarify that the winpe itself seems to be just fine. It recognize my NIC and disk. I can access the iSCSI disk in the cmd window in the winpe. I can redirect dir output to the disk and use type command to read it back.

What appears to go wrong is the 22H2 setup.exe itself. But note I'm saying setup.exe running inside the winpe environment, just in parallel with my cmd window. I assume it's sharing the winpe driver that works just fine in the cmd window. The setup.exe is able to partition the disk -- since this change persist across boot. But it fails to copy files to the disk. This time I attempt to check the change the setup.exe made to the disk after crash reboot. It seems that the situation varies -- sometimes the disk is still empty, some times I find $WINDOWS.BT directory with some non-empty files (including the setupact.log) -- but the content of those files are weird (looks like white spaces, didn't check with hex-editors). Maybe the contents in those files are lost due to crash reboot. Not sure if I'll be lucky enough to be able to capture some error log ...

As another desperate attempt, I tried my luck with an 1809 ISO, which is known to work according to the link I post in at the top of this thread. It has went past the crash reboot point that 22H2 suffers. Hopefully this can run to finish.

But what if it tries to upgrade to 22H2 after install and crash once again?

**Reply by NiKiZe** [reply]
Is your installation clean? I assume that you at least most of the time reach the partition part of the setup process?

**Reply by firemeteorxx** [reply]
Reaching partition part is guaranteed. Crashing reboot happens afterwards when it performs file copying task.
I can't make it to the completion of the installation, if that's what you mean by clean install. 

BTW, I am installing to an empty disk, instead of doing upgrade from a bootable disk. But doing windows upgrade to 22H2 also results in failure && rollback as I mentioned before. I can collect setup log in this case, but there are actually too many errors to figure out the fatal one...

**Comment by firemeteorxx** [comment]
Just to report back, the 1809 ISO can install to iSCSI disk and san-boot just fine. Both NICs and disks are doing good. After the new windows configured the network, it immediately offers to download the new 22H2 release, which frighten me to death :-)

**Comment by firemeteorxx** [comment]
Tried another time to upgrade to 22H2, this time from the 1809 installation I managed to achieve in last post.
Unfortunately it's still a failure. The problem for the failed upgrade appears to be that the installer failed to boot after restart.
The installer finally droped me into a recovery environment. I wasn't able to find any disk nor any NIC inside the cmd window.
I suspect this specific feature upgrade installer does not carry the driver for my NIC. 
Yet it does not allow me to inject a driver in any way.

Anything I can do here to make progress?

**Reply by NiKiZe** [reply]
Not sure if ms officially even supports iSCSI boot on win10, other than that make sure you use a nic/driver that officially supports it, and then contact that vendor.

So answer is "try a different NIC"
Not much we can do here.

**Reply by dfh0522** [reply]
> Anything I can do here to make progress?

Here are my experiences with iSCSI boot Windows 10 & 11, but this is only for BIOS not for UEFI.
The chained ipxe.conf:

:bm1adu
dhcp
set iscsi-rootfs iscsi:192.168.2.11::3262::iqn.2003-01.org.linux-iscsi.b85m.x8664:bm1adu
#set keep-san 1
sanboot --drive 0x80 --keep ${iscsi-rootfs} || shell

:win1022h2
set iscsi-rootfs iscsi:192.168.2.11::3262::iqn.2003-01.org.linux-iscsi.b85m.x8664:bm1adu
sanhook --drive 0x80 ${iscsi-rootfs}

kernel tftp://192.168.2.11/wimboot
initrd tftp://192.168.2.11/win1022h2/bcd         BCD
initrd tftp://192.168.2.11/win1022h2/boot.sdi    boot.sdi
initrd tftp://192.168.2.11/win1022h2/boot.wim    boot.wim
boot || shell

:win1122h2
set iscsi-rootfs iscsi:192.168.2.11::3262::iqn.2003-01.org.linux-iscsi.b85m.x8664:bm1adu
sanhook --drive 0x80 ${iscsi-rootfs}

kernel tftp://192.168.2.11/wimboot
initrd tftp://192.168.2.11/win1122h2/bcd         BCD
initrd tftp://192.168.2.11/win1122h2/boot.sdi    boot.sdi
initrd tftp://192.168.2.11/win11ent/boot.wim    boot.wim
boot || shell

Then follow the steps found from Internet:
1. Shift+F10 opens commandline in Windows installator
2. Wpeutil InitializeNetwork
3. Wpeutil DisableFirewall
4. setup.exe with /noreboot
5. regedit to hive SYSTEM and disable Paging, just clean up the string of Paging file registry. 
6. Wpeutil shutdown

I need to change the target of ipx.conf from win1022h2 to bm1adu before 2nd boot.
These work for me with win1021h2 & win1022h2, except for win1122h2, it needs to copy bootmgr from 1st patition of win10 to override win11 iSCSI drive twice, one for 2nd boot installation, another for system driver upgrade during Windows update.

**Reply by firemeteorxx** [reply]
Thanks for your information, dfh0522!
It looks like a pretty standard setup beside the paging disable trick that I also came across in different threads.
But I'm afraid I haven't get to the paging specific failure yet, it fails much earlier at my site.
I can retry later to see if your step #2 disabling the firewall helps, don't have high confidence to be honest...

So this indeed looks like an issue specific to my environment, unfortunately.
@dfh0522, would you mind sharing the HW you are using? I guess the NIC should matter most.

BTW, did you have old Win10 machine that upgrades automatically and successfully to W10 2022H2?
I suffered problem with my 1709 box. 
I wonder if there is anything I can do before it reboots into the upgrading process, e.g. to inject some NIC drivers etc.
Have you heard of any story in the Internet about this kind of hack?

**Reply by dfh0522** [reply]
The machines are installed with TP-Link TG-3468(realtek) & HP 561FLR-T(intel), I don't have old version ISO, but it's no problem to upgrade from 20h2 to 22h2.

**Reply by firemeteorxx** [reply]
It has been a long time, but I picked this up recently.

Turns out that for win10 22h2 setup program to proceed, the following line from @dfh0522 is crucial. 
> Wpeutil DisableFirewall

This line helps me move beyond the crash reboot point that bothered me a long time.
> regedit to hive SYSTEM and disable Paging, just clean up the string of Paging file registry.

This line is still relevant for win10 22h2 and it solves the DRIVER_IRQL_NOT_LESS_OR_EQUAL error during boot stage.

Thank you for the information.

However, I'm still not able survive the 1st reboot, at least in my virtual-box VM with 82540EM NIC. This time I'm stuck in the recovery screen, complaining errors code 0xc0000225 from winload.exe.

I suspect this time it's NIC driver related. But my NIC should be supported out-of-box in both v1709 and in 22h2 installer PE.
Actually I attempted driver injection with DISM for my I-218V && I-219V NIC, but the new driver is not used by the PE at all.
If the built-in driver is actually the offender, I'll need to figure out a way to remove them first...

How do people diagnose windows boot issues like this? 
The error info is mystery and very little relevant info can be found in the search engine...

**Comment by PatcoreShare** [comment]
Cześć, mam taki sam proble. 
Komputer Dell Service Tag DKZPQR3.
Przy pierwszym uruchomieniu ponownie "CRITICAL_PROCESS_DIED" 

WinPE - Upgrade and install drivers
Próboowałe ze strony dell 
Próboowałem ze strony https://www.intel.com/content/www/us/en/products/sku/82185/intel-ethernet-connection-i219lm/ordering.html?erpm_id=12789487_ts1672226145945

Any suggestions on how to fix it?

**Comment by epacua** [comment]
I have successfully installed win10-22h1 and h2 several times. That sounds like that your install.wim does not have all the nic drivers and/or iscsi components. You'd have to embed those via dism.exe.

What does the log at iscsi target says?

**Reply by PatcoreShare** [reply]
Am I doing this correctly? 
1. NIC driver
![image](https://user-images.githubusercontent.com/11398415/214227097-c1d1cc2b-66e1-4d91-8c17-81929e18a82c.png)
`dism /image:"d:\winPE" /add-driver /driver:"D:\NIC\0\Drivers\Dell-I219-12.19.1.37\e1d68x64.inf"`

2. Update WinPE
![image](https://user-images.githubusercontent.com/11398415/214229033-758ec0de-ba6f-4e62-b09b-9fe93968cea8.png)

`Dism /Add-Package /Image:"D:\ winPE " /PackagePath:"D:\update\windows10.0-kb5019275-x64_00d644d15b7d944e71fd1dba75ed5a06c1107757.msu"`


4. Add iSCSI components
![image](https://user-images.githubusercontent.com/11398415/214228627-4af9b7ca-d35c-47cb-82bf-5ea60242f31c.png)

`Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-WMI.cab"`

`Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-NetFx.cab"`

`Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-Scripting.cab"`

`Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-PowerShell.cab"`

`Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-StorageWMI.cab"`

Am I forgot something?

**Reply by epacua** [reply]
> Am I doing this correctly?
> 
> 1. NIC driver
>    ![image](https://user-images.githubusercontent.com/11398415/214227097-c1d1cc2b-66e1-4d91-8c17-81929e18a82c.png)
>    `dism /image:"d:\winPE" /add-driver /driver:"D:\NIC\0\Drivers\Dell-I219-12.19.1.37\e1d68x64.inf"`
> 2. Update WinPE
>    ![image](https://user-images.githubusercontent.com/11398415/214229033-758ec0de-ba6f-4e62-b09b-9fe93968cea8.png)
> 
> `Dism /Add-Package /Image:"D:\ winPE " /PackagePath:"D:\update\windows10.0-kb5019275-x64_00d644d15b7d944e71fd1dba75ed5a06c1107757.msu"`
> 
> 4. Add iSCSI components
>    ![image](https://user-images.githubusercontent.com/11398415/214228627-4af9b7ca-d35c-47cb-82bf-5ea60242f31c.png)
> 
> `Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-WMI.cab"`
> 
> `Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-NetFx.cab"`
> 
> `Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-Scripting.cab"`
> 
> `Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-PowerShell.cab"`
> 
> `Dism /Add-Package /Image:"d:\ winPE " /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-StorageWMI.cab"`
> 
> Am I forgot something?

The steps themselves look good to me. But you'd have to do the same steps for both boot.wim (WinPE) and more importantly to install.wim (the actual Win10 installer).

The way I install Win10 is I pxe boot onto WinPE via ipxe, then mount the smb share where the win10 installer is located. Then I confirm that the iscsi lun is detected via diskpart. Finally, install from there.

**Reply by PatcoreShare** [reply]
> The way I install Win10 is I pxe boot onto WinPE via ipxe, then mount the smb share where the win10 installer is located. Then I confirm that the iscsi lun is detected via diskpart. Finally, install from there.

I do the same as you described.

>But you'd have to do the same steps for both boot.wim (WinPE)
It work

>and more importantly to install.wim (the actual Win10 installer).


`dism /delete-image /imagefile:d:\install.wim /index:X`
`dism /get-imageinfo /imagefile:d:\install.wim `
![image](https://user-images.githubusercontent.com/11398415/214304829-7cfd3a01-8926-40a6-9e0d-f787ff9fa78c.png)

I have problems with this :
```
Dism /Add-Package /Image:"d:\Winpro" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-WMI.cab"

Dism /Add-Package /Image:"d:\Winpro" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-NetFx.cab"

Dism /Add-Package /Image:"d:\Winpro" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-Scripting.cab"

Dism /Add-Package /Image:"d:\Winpro" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-PowerShell.cab"

Dism /Add-Package /Image:"d:\Winpro" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-StorageWMI.cab"
```

![image](https://user-images.githubusercontent.com/11398415/214305785-d51da612-cca7-426f-b2bf-6ce429ef69db.png)

I'm out of ideas.

**Comment by epacua** [comment]
@patriocoz Here's the exact script I used when rebuilding wim file. You forgot to copy the iscsi-related files from host into the mounted image.

`set WindowsPE=C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs     
if not exist C:\Mounted_WIM md C:\Mounted_WIM   
Dism /Mount-Image /ImageFile:C:\WinPE\media\sources\boot.wim /Index:1 /MountDir:C:\Mounted_WIM   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\WinPE-WMI.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\en-us\WinPE-WMI_en-us.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\WinPE-NetFX.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\en-us\WinPE-NetFX_en-us.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\WinPE-Scripting.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\en-us\WinPE-Scripting_en-us.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\WinPE-PowerShell.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\en-us\WinPE-PowerSHell_en-us.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\WinPE-StorageWMI.cab"   
Dism /Add-Package /Image:C:\Mounted_WIM /PackagePath:"%WindowsPE%\en-us\WinPE-StorageWMI_en-us.cab"   
copy C:\windows\system32\iscsicpl.dll C:\Mounted_WIM\windows\system32   
copy C:\windows\system32\iscsicpl.exe C:\Mounted_WIM\windows\system32   
copy C:\windows\system32\en-us\iscsicpl.dll.mui C:\Mounted_WIM\windows\system32\en-us   
copy C:\windows\system32\en-us\iscsicpl.exe.mui C:\Mounted_WIM\windows\system32\en-us                                                                                                            
Dism /Get-Packages /Image:C:\Mounted_WIM   
pause   
Dism /Unmount-Image /MountDir:C:\Mounted_WIM /commit  

`

**Reply by PatcoreShare** [reply]
I didn't know about this:
```
copy C:\windows\system32\iscsicpl.dll C:\Mounted_WIM\windows\system32
copy C:\windows\system32\iscsicpl.exe C:\Mounted_WIM\windows\system32
copy C:\windows\system32\en-us\iscsicpl.dll.mui C:\Mounted_WIM\windows\system32\en-us
copy C:\windows\system32\en-us\iscsicpl.exe.mui C:\Mounted_WIM\windows\system32\en-us
```

Are you copy this to boot.wim (WinPE), boot.wim (Windows10) or install.wim (Windows 10)

**Reply by epacua** [reply]
@patriocoz You'd have to copy those to both boot and install.wim.

**Reply by firemeteorxx** [reply]
I wonder why we need those iSCSI initiator capability in the context of iPXE boot? I can see how it would be helpful with a USB boot since the iSCSI disk won't be made available without an built-in initiator. But with iPXE the sanhook feature should setup the iBFT and the OS setup media should be able to recognize without problem. At least in my case my problem is likely somewhere else.

**Reply by epacua** [reply]
If you didn't embed the iscsi components on both boot.wim and install.wim, then that is the most likely cause of your problem. Else, it could be driver issue on either or both ipxe and *.wim.

![image](https://github.com/ipxe/ipxe/assets/15176373/db6786d1-b23b-4f8d-b45c-cd1359f2c41a)

**Reply by firemeteorxx** [reply]
I'm glad to report that I was finally able to install windows 10 22h2 version on my box with I-218V NIC.

The only tweak I made to make it work was this line:
> Wpeutil DisableFirewall

It used to work without this line up to 1809 version. Apparently something has changed since...

It still remains an issue for my Virtualbox VM due to 82540EM NIC does not survive the 1st reboot after the PE part is over.


For my I-218V NIC box, I would add the following:

I did tried to inject NIC drivers through `DISM /add-driver` to both of my boot.wim && install.wim. But from what I was able to figure out in the registry, at least the WinPE part of the installation process was not using the newly injected e1d driver.

The install media I used was the stock Microsoft ISO except for the NIC driver injection mod I attempted. I may have made mistakes. But if I read the `dsim /get-packages` result correctly, there is no StorageWMI. There is also no iscsicpl.exe in the mounted WIM image. I think the iBFT prepared by the ipxe sanhook is doing the magic for me.

**Reply by epacua** [reply]
With regards to Virtualbox, the builtin iPXE there is castrated, i.e, it is not capable of sanbooting. You can confirm it by checking the displayed message. There were discussions around it by compiling your own iPXE rom then inject it, but there is a size limitation of <=57k. I tried that before but failed miserably. I would be interested if you would be able to compile one.

**Reply by firemeteorxx** [reply]
In my case I use iPXE chain loading for all my boxes. That's why I'm not aware of the limitation you mentioned. 

Building the image shouldn't be hard -- any Linux box with DEV env should do. But squeezing it to fit the 57K limitation might not be trivial. At least none of the images I have can fit. iPXE comes with feature tailoring capability. So this might work if you carefully cherry-pick only the features you plan to use.

BTW1: I'm curious where does this 57K limitation come from?
BTW2: It seems pointless if the VirtualBox emulated NICs does not survive the Windows boot process? Or do you happen to know a trick?
BTW3: I also tried to play around the ipxe UEFI boot flow. It seems that Virtualbox VM can be configured to use UEFI instead of BIOS. The UEFI firmware in the VM comes with iSCSI initiator and PXE features. But so far I was not able to get anything out of those features... The PXE boot process appears to hang without any diagnose info. The iSCSI initiator feature could be configured but I was not able to figure out how to get it show up in the boot menu...

**Reply by epacua** [reply]
It was ages ago when I last tried to sanboot with VirtualBox. For now, I stick with either KVM or Xen.

If you check the ipxe source code, there is a profile there called vbox, see below:

```
ipxe/src/config/vbox]$ grep -A2 '^Max size' README 
Max size of a VirtualBox ROM is 56KB, 57344 bytes. There should be no need
to pad the image as long as the binary is smaller or equal to this size.
```

In the past, I went to great lengths like disabling options that I thought I'd not need. I tried the one below, but still no luck.

`
make CONFIG=vbox bin/intel--virtio-net--pcnet32.isarom
`

I think I tried chainloading a non-castrated undionly.kpxe as well but without much success. I had even better success with an old Mellanox ConnectX-2 with gPXE on it which is extremely ancient. I even went to flash that with kinda newer ipxe rom.

With UEFI, I had much better success with KVM. I use ipxe.efi.

**Reply by firemeteorxx** [reply]
> I did tried to inject NIC drivers through DISM /add-driver to both of my boot.wim && install.wim. But from what I was able to figure out in the registry, at least the WinPE part of the installation process was not using the newly injected e1d driver.

I may have to correct this comment according to my experiment on the old 1709 box.
The NIC driver for I-219V card in my system:
1. 1709 version: e1i63x64.sys
2. 22h2 installer WinPE: e1i65x64.sys
3. 22h2 fresh install: e1d.sys (the one I injected into the WIM).

I tried to upgrade my NIC driver in my 1709 box. My first attempt through auto search ends up with e1i65x64.sys, which leads to an INACCESSIBLE_BOOT_DEVICE BSOD. I manually checked this driver before reboot with bindview.exe from WDK, the WPF setting appears to be correct. But still it was not able to boot for unknown reason. Maybe not in the boot-critical driver DB?

My second attempt is to manually specify the e1d.sys driver which is known to work on 22h2. And magically this one survives the reboot without any issue.  Unfortunately, with this new driver I still cannot go through the 22h2 upgrade process. I suspect the upgrade release still uses the e1i65x64.sys which is buggy in my box... I'm not sure if there is anything I can do to get e1i65x64.sys work in my 1709, but as long as the upgrade process uses its own NIC driver this likely won't help the upgrade process at all...

> It was ages ago when I last tried to sanboot with VirtualBox. For now, I stick with either KVM or Xen.

I use both VBox and Xen, but has little experience with KVM. I use XEN to host VMs on my home server and VBox to host VMs on my desktop, mostly focus only Windows in the latter  scenario. Does KVM have an easy to use GUI app like VBox?

I noticed that in your screenshot (KVM based I assume?) the NIC is named as `Intel PRO/1000MT Network Connection`. The VBox emulated NIC 82540EM was called `Intel PRO/1000 MT Desktop Adapter` which is pretty similar to yours but could not survive the installation. Could you share the NIC driver and PCI ID in your VM?

>I think I tried chainloading a non-castrated undionly.kpxe as well but without much success.

It's true that the mileage highly depends on the NIC and the driver (iPXE) version. Currently iPXE 1.21.1 happens to work. But I was not always as lucky in the past...

> make CONFIG=vbox bin/intel--virtio-net--pcnet32.isarom

This virtio-net NIC could be interesting. I never considered this before. But does it really work with windows VM?

**Reply by epacua** [reply]
Here are screen shots showing the device and driver IDs for the Intel NIC.

![win10-vm](https://github.com/ipxe/ipxe/assets/15176373/2741c1c2-5abb-43be-b807-b89ef2fec4d1)

![win10-vm2](https://github.com/ipxe/ipxe/assets/15176373/d5cbbc5a-b512-4481-9ba8-f9053f6647df)

Yes, my vm is hosted on KVM (though you may notice that the iscsi drive is called 'xen03-whatever'. It's just the name of iscsi lun).

Here's the xml for that. Notice that it has not got any hard drive at all. Also, it boots from network. On line# 30, that's where you decide which nic model you want to emulate.


```
[user@pc1:~]$ cat -n xen03-win10.xml 
     1  <domain type='kvm'>
     2          <name>xen03-win10</name>
     3      <uuid>28a6cb38-fc57-40f6-89d4-b0f8fa366780</uuid>
     4    <memory unit='GiB'>8</memory>
     5    <currentMemory unit='GiB'>8</currentMemory>
     6    <vcpu placement='static'>4</vcpu>
     7    <os>
     8      <type arch='x86_64' machine='q35'>hvm</type>
     9    </os>
    10    <features>
    11      <acpi/>
    12      <apic/>
    13      <viridian/>
    14      <hyperv/>
    15    </features>
    16    <clock sync='localtime'/>
    17    <on_poweroff>destroy</on_poweroff>
    18    <on_reboot>restart</on_reboot>
    19    <on_crash>restart</on_crash>
    20
    21          <devices>
    22                  <tpm model='tpm-crb'>
    23                          <backend type='emulator' version='2.0'/>
    24                  </tpm>
    25      <emulator>/usr/bin/qemu-system-x86_64</emulator>
    26
    27      <interface type='network'>
    28        <mac address='00:16:3e:4e:f9:03'/>
    29        <source network='xen-virbr10'/>
    30        <model type='e1000'/>
    31                   <boot order='1'/>
    32      </interface>
    33      <serial type='pty'>
    34        <target port='0'/>
    35      </serial>
    36      <console type='pty'>
    37        <target type='serial' port='0'/>
    38      </console>
    39      <input type='tablet' bus='usb'/>
    40      <input type='mouse' bus='ps2'/>
    41      <input type='keyboard' bus='ps2'/>
    42
    43      <graphics type='spice' port='2050' autoport='no' listen='127.0.0.1'>
    44        <listen type='address' address='127.0.0.1'/>
    45        <mouse mode='server'/>
    46        <clipboard copypaste='no'/>
    47      </graphics>
    48      <video>
    49        <model type='qxl' ram='131072' vram='131072' vgamem='131072' heads='1' primary='yes'/>
    50        <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
    51      </video>
    52      <sound model='ich9'>
    53        <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    54      </sound>
    55      <audio id='1' type='spice'>
    56        <output name='kali' latency='100' mixingEngine='yes' fixedSettings='yes' bufferLength='128'>
    57          <settings frequency='44100' channels='2' format='f32'/>
    58        </output>
    59      </audio>
    60
    61      <memballoon model="virtio">
    62        <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
    63      </memballoon>
    64          </devices>
    65  </domain>
```

In regards to vbox, I pretty much gave up on it. Though have to admit that VirtualBox GUI is the best (I've used vmware workstation at work and am not a big fan of their clunky UI). On linux side, you can use virt-viewer, remmina or remote-viewer. All of them has their own level of suckiness. VBox UI beat them all.

By the way, you still(?) haven't mentioned the following:
- ipxe bootrom used (I stick to custom undionly.kpxe and ipxe.efi for uefi)
- You mentioned you're using Xen, care to share your .cfg file? Or if using through libvirt, then must be xml.
- iscsi target. I use target hosted on AlmaLinux 9.2
- dhcp server. I use Kea
- have you done any packet capture either through wireshark or tcpdump when troubleshooting?

Hope this helps.

**Reply by firemeteorxx** [reply]
Thanks for the info on your working emulated NIC. From what I can tell, the difference between the 82540EM in VBox and yours in KVM is lies in the SUBSYS and the REV part. Given both NIC share the same driver, I'm not sure if I should expect the difference in the NIC model being responsible to the difference in the result.
![image](https://github.com/ipxe/ipxe/assets/6914654/ddd2470f-2624-4da5-8db0-3271cda797f6)

So far I've been focusing on the NIC and the windows itself. But now that you asked, here are those info for reference:
> ipxe bootrom used (I stick to custom undionly.kpxe and ipxe.efi for uefi)

I'm on iPXE 1.21.1 with kpxe && efi flavors. I tend to stick to dedicated NIC drivers since I experienced weirdness related to undi driver before. 
bin-x86_64-efi/intel--realtek--hermon--virtio-net.efi
bin/intel--realtek--hermon--virtio-net--undi.kpxe

> You mentioned you're using Xen, care to share your .cfg file? Or if using through libvirt, then must be xml.

XEN is a distraction here since the VM that I'm struggling at is VBox based.
The XEN VM that I use everyday is currently NFS based. I'm not sure iSCSI will be better than NFS.
Maybe I can give it a try later...

> iscsi target. I use target hosted on AlmaLinux 9.2

FreeNAS VM running on-top of XEN. An all-in-one build. The NAS VM serves storage to everything else.
> dhcp server. I use Kea

Dnsmasq on openwrt.
> have you done any packet capture either through wireshark or tcpdump when troubleshooting?

Not for this discussion. I did use Wireshark to debug the EFI boot setup and finally got it work. But for this case, I still suspect NIC driver issue in Windows.

Actually the VM in my discussion is mostly an experiment vehicle, even though it ends up with leading to more noises than observations. The main scenario that I would need to rely on iSCSI boot is running Windows on my physical boxes (the ones with I218V && I219V respectively). So far I'm able to get Win10 22h2 running on both of them. The major concern I have is if a future upgrade will break the current windows install once again. I know Win10 is approaching EOL and there won't be any major upgrade. But when various SWs begin to move away from it, I may need to suffer once again if the process is not smooth at all.
